arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/ops/SwitchMapTest.kt

```diff
@@ -0,0 +1,163 @@
+package arrow.fx.coroutines.stream.ops
+
+import arrow.fx.coroutines.Atomic
+import arrow.fx.coroutines.Promise
+import arrow.fx.coroutines.Semaphore
+import arrow.fx.coroutines.StreamSpec
+import arrow.fx.coroutines.assertThrowable
+import arrow.fx.coroutines.milliseconds
+import arrow.fx.coroutines.sleep
+import arrow.fx.coroutines.stream.Stream
+import arrow.fx.coroutines.stream.append
+import arrow.fx.coroutines.stream.compile
+import arrow.fx.coroutines.stream.terminateOnNull
+import arrow.fx.coroutines.throwable
+import arrow.fx.coroutines.unit
+import io.kotest.matchers.shouldBe
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.int
+
+class SwitchMapTest : StreamSpec(iterations = 250, spec = {
+  "flatMap equivalence when switching never occurs" {
+    checkAll(Arb.stream(Arb.int())) { s ->
+      val expected = s.compile().toList()
+      Stream
+        .effect { Semaphore(1) }
+        .flatMap { guard ->
+          // wait for inner to emit to prevent switching
+          s.effectTap { guard.acquire() }
+            // outer terminates, wait for last inner to emit
+            .onFinalize { guard.acquire() }
+            .switchMap { x -> Stream.emits(x).onFinalize { guard.release() } }
+        }
+        .compile()
+        .toList() shouldBe expected
+    }
+  }
+
+  "inner stream finalizer always runs before switching" {
+    checkAll(Arb.stream(Arb.int())) { s ->
+      Stream.effect {
+        val ref = Atomic(true)
+        s.switchMap {
+          Stream.effect {
+            ref.get()
+          }.flatMap { released ->
+            if (!released) Stream.raiseError(Throwable())
+            else
+              Stream.effect {
+                ref.set(false)
+                sleep(20.milliseconds)
+              }
+                .onFinalize {
+                  sleep(100.milliseconds)
+                  ref.set(true)
+                }
+          }
+        }
+      }
+        .compile()
+        .drain()
+    }
+  }
+
+  "when primary stream terminates, inner stream continues"  {
+    checkAll(Arb.stream(Arb.int()), Arb.stream(Arb.int())) { s1, s2 ->
+      val expected = s1.last().terminateOnNull().flatMap { s -> s2.append { Stream(s) } }.compile().toList()
+
+      s1.switchMap { s -> Stream.sleep_(20.milliseconds).append { s2 }.append { Stream(s) } }
+        .compile()
+        .toList() shouldBe expected
+    }
+  }
+
+  "when inner stream fails, overall stream fails"  {
+    checkAll(Arb.stream(Arb.int()), Arb.throwable()) { s0, e ->
+      val s = Stream(0).append { s0 }
+        .delayBy(25.milliseconds)
+        .flatMap { Stream.raiseError<Int>(e) }
+        .compile()
+
+      assertThrowable {
+        s.drain()
+      } shouldBe e
+    }
+  }
+
+  "when primary stream fails, overall stream fails and inner stream is terminated" {
+    checkAll(Arb.throwable()) { e ->
+      val s = Stream.effect { Semaphore(0) }
```

Comment:
Nit: these can also be created outside of the `Stream` since the tests are in `suspend` context.
This requires less nesting of `flatMap` of `Stream` in code.